src/Symfony/Component/Uri/Uri.php

```diff
@@ -0,0 +1,172 @@
+ */
+final class Uri implements \Stringable
+{
+    private const URI_GLOBAL_REGEX = '/^(?:(?P<scheme>[^:\/?#]+):)?(?:\/\/(?P<authority>[^\/?#]*))?(?P<path>[^?#]*)(?:\?(?P<query>[^#]*))?(?:#(?P<fragment>.*))?$/';
+    private const URI_AUTHORITY_REGEX = '/^(?:(?P<user>[^:@]*)(?::(?P<pass>[^@]*))?@)?(?P<host>[^:]*)(?::(?P<port>\d*))?$/';
+
+    public function __construct(
+        public string $scheme,
+        #[\SensitiveParameter]
+        public ?string $user = null,
+        #[\SensitiveParameter]
+        public ?string $password = null,
+        public ?string $host = null,
+        public ?int $port = null,
+        public ?string $path = null,
+        public ?QueryString $query = null,
+        public ?string $fragment = null,
+        public ?FragmentTextDirective $fragmentTextDirective = null,
+    ) {
+    }
+
+    /**
+     * Parses a URL.
+     *
+     * The `user` and `pass` keys are url-decoded automatically when parsing.
+     *
+     * @throws InvalidUriException
+     */
+    public static function parse(string $uri): static
+    {
+        preg_match(self::URI_GLOBAL_REGEX, $uri, $matches);
+        if (!$matches || !isset($matches['scheme']) || '' === $matches['scheme']) {
+            throw new InvalidUriException($uri);
+        }
+
+        if (isset($matches['authority'])) {
+            preg_match(self::URI_AUTHORITY_REGEX, $matches['authority'], $authMatches);
+
+            $matches = array_merge($matches, $authMatches);
+            unset($matches['authority']);
+        }
+
+        $matches = array_filter($matches, static fn (string $value): bool => '' !== $value);
+
+        return new static(
+            $matches['scheme'],
+            isset($matches['user']) ? rawurldecode($matches['user']) : null,
+            isset($matches['pass']) ? rawurldecode($matches['pass']) : null,
+            $matches['host'] ?? null,
+            $matches['port'] ?? null,
+            $matches['path'] ?? null,
+            isset($matches['query']) ? QueryString::parse($matches['query']) : null,
+            $matches['fragment'] ?? null,
+        );
+    }
+
+    /**
+     * Resolves a relative URI against a base URI.
+     *
+     * Uri::resolve('http://example.com', '/foo/bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', '../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '../../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', 'http://example.org/bar'); // http://example.org/bar
+     */
+    public static function resolve(self|string $baseUri, string $relativeUri): string
```

Comment:
Should this method returns a Uri instance ?

Reply:
Good question, we went for a string while specifying the component. However, why not? I'm not sure what people are expecting to do with the result of such of method. Is a string enough or should we parse it once resolved? I don't have a strong opinion here 🤔

Reply:
Me neither :) 

As you have an Uri instance just before returning the string result... it kinda made sense to me.

.... but as the method allow 2 strings in arguments, this can also be non-obvious.

🤷‍♂️